File: banka-frontend-enhanced/src/lib/permissions.test.ts
import { describe, expect, it } from 'vitest'
import { Permissions, has, hasAny } from './permissions'

describe('has', () => {
  it('matches when the holder has the exact permission', () => {
    expect(has([Permissions.EmployeeRead], Permissions.EmployeeRead)).toBe(true)
  })

  it('returns false when the holder lacks the permission', () => {
    expect(has([Permissions.ClientRead], Permissions.EmployeeWrite)).toBe(false)
  })

  it('treats Admin as a wildcard', () => {
    expect(has([Permissions.Admin], Permissions.EmployeeWrite)).toBe(true)
    expect(has([Permissions.Admin], Permissions.PermissionGrant)).toBe(true)
  })

  it('handles an empty holder list', () => {
    expect(has([], Permissions.EmployeeRead)).toBe(false)
  })
})

describe('hasAny', () => {
  it('returns true on any match', () => {
    expect(
      hasAny([Permissions.ClientRead], [Permissions.EmployeeWrite, Permissions.ClientRead]),
    ).toBe(true)
  })

  it('returns false when nothing matches', () => {
    expect(
      hasAny([Permissions.ClientRead], [Permissions.EmployeeWrite, Permissions.PermissionGrant]),
    ).toBe(false)
  })

  it('still treats Admin as wildcard', () => {
    expect(hasAny([Permissions.Admin], [Permissions.EmployeeWrite])).toBe(true)
  })

  it('returns false for an empty target list', () => {
    expect(hasAny([Permissions.Admin], [])).toBe(false)
  })
})

File: banka-frontend-enhanced/Banka-3-Backend-rewrite/pkg/permissions/permissions_test.go
package permissions

import "testing"

func TestHas(t *testing.T) {
	holder := []string{Admin, EmployeeRead, ClientRead}
	if !Has(holder, Admin) {
		t.Error("Has should find Admin")
	}
	if Has(holder, EmployeeWrite) {
		t.Error("Has should not find EmployeeWrite")
	}
	if Has(nil, Admin) {
		t.Error("Has on nil should be false")
	}
	if Has([]string{}, Admin) {
		t.Error("Has on empty should be false")
	}
}

func TestHasAny(t *testing.T) {
	holder := []string{EmployeeRead}
	if !HasAny(holder, Admin, EmployeeRead) {
		t.Error("HasAny should match EmployeeRead")
	}
	if HasAny(holder, Admin, EmployeeWrite) {
		t.Error("HasAny should not match anything")
	}
	if HasAny(nil, Admin) {
		t.Error("HasAny on nil holder should be false")
	}
	if HasAny(holder /* no targets */) {
		t.Error("HasAny with no targets should be false")
	}
}

func TestHasAll(t *testing.T) {
	holder := []string{Admin, EmployeeRead, EmployeeWrite}
	if !HasAll(holder, Admin, EmployeeRead) {
		t.Error("HasAll should match a subset")
	}
	if HasAll(holder, Admin, ClientWrite) {
		t.Error("HasAll should fail on missing target")
	}
	if !HasAll(holder /* no targets */) {
		t.Error("HasAll with no targets should be vacuously true")
	}
}

func TestRoleBundlesContent(t *testing.T) {
	// Lock down the c1 role bundles so an accidental rename or reorder
	// surfaces as a test failure rather than a runtime change.
	cases := []struct {
		name   string
		bundle []string
		want   []string
	}{
		{"client-basic", RoleClientBasic, []string{ClientRead, AccountRead, CardRead, CardWrite, PaymentWrite, LoanRead, LoanWrite}},
		{"employee-basic", RoleEmployeeBasic, []string{EmployeeRead, ClientRead, AccountRead, CompanyRead, CardRead, LoanRead}},
		{"employee-agent", RoleEmployeeAgent, []string{ClientRead, ClientWrite, AccountRead, AccountWrite, CompanyRead, CompanyWrite, CardRead, CardWrite, LoanRead, LoanWrite}},
		{"employee-supervisor", RoleEmployeeSupervisor, []string{EmployeeRead, ClientRead, ClientWrite, AccountRead, AccountWrite, CompanyRead, CompanyWrite, CardRead, CardWrite, LoanRead, LoanWrite}},
		{"employee-admin", RoleEmployeeAdmin, []string{Admin, EmployeeRead, EmployeeWrite, ClientRead, ClientWrite, CompanyRead, CompanyWrite, AccountRead, AccountWrite, CardRead, CardWrite, LoanRead, LoanWrite, PaymentWrite, ExchangeWrite, PermissionGrant, Actuary, ActuarySupervisor, TradingMargin}},

		{"client-trading", RoleClientTrading, []string{
			ClientRead, AccountRead, CardRead, CardWrite, PaymentWrite, LoanRead, LoanWrite,
			// Spec p.4: stocks + OTC + funds are one capability, gated
			// by the single TradingClient flag.
			TradingClient,
		}},
		{"actuary-supervisor", RoleEmployeeActuarySupervisor, []string{
			Actuary, ActuarySupervisor, TradingMargin, EmployeeRead,
			// c4 additions — supervisors run OTC, manage funds, see Profit Banke.
			OTCTradeSupervisor,
			FundsReadSupervisor, FundsManageSupervisor,
			BankProfitRead,
		}},
		{"actuary-agent", RoleEmployeeActuaryAgent, []string{Actuary, ActuaryAgent}},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if len(tc.bundle) != len(tc.want) {
				t.Fatalf("len mismatch: got %v want %v", tc.bundle, tc.want)
			}
			for i := range tc.want {
				if tc.bundle[i] != tc.want[i] {
					t.Fatalf("[%d] got %q want %q", i, tc.bundle[i], tc.want[i])
				}
			}
		})
	}
}
